src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpCodeVisitor.cs

```diff
@@ -581,6 +561,31 @@ private CSharpCodeWriter RenderPreWriteStart()
             return RenderPreWriteStart(Writer, Context);
         }
 
+        private void RenderChunkedWriteLiteral(string text)
+        {
+            var charactersRendered = 0;
+
+            // Render the string in pieces to avoid Roslyn OOM exceptions at compile time:
+            // https://github.com/dotnet/roslyn/issues/7398
+            while (charactersRendered < text.Length)
+            {
+                if (Context.ExpressionRenderingMode == ExpressionRenderingMode.WriteToOutput)
+                {
+                    RenderPreWriteStart();
+                }
+
+                var charactersToSubstring = Math.Min(MaxStringLiteralLength, text.Length - charactersRendered);
+                var textToRender = text.Substring(charactersRendered, charactersToSubstring);
+                Writer.WriteStringLiteral(textToRender);
```

Comment:
Ick.
1. Don't `Substring()` if `text.Length <= MaxStringLiteralLength`.
2. Otherwise call `text.ToCharArray()` once before the loop and pass the array, start index, and length into the `Writer`.
3. Expose the necessary wrapper methods in `CodeWriter` (`WriteStringLiteral()`) and `CSharpCodeWriter` (`Write()`). Those APIs are anemic compared to the underlying `StringWriter` and lousy for this use case.